compiler/rustc_mir/src/transform/abort_unwinding_calls.rs

```diff
@@ -0,0 +1,107 @@
+use crate::transform::MirPass;
+use rustc_middle::middle::codegen_fn_attrs::CodegenFnAttrFlags;
+use rustc_middle::mir::*;
+use rustc_middle::ty::layout;
+use rustc_middle::ty::{self, TyCtxt};
+use rustc_target::spec::abi::Abi;
+
+/// A pass that runs which is targeted at ensuring that codegen guarantees about
+/// unwinding are upheld for compilations of panic=abort programs.
+///
+/// When compiling with panic=abort codegen backends generally want to assume
+/// that all Rust-defined functions do not unwind, and it's UB if they actually
+/// do unwind. Foreign functions, however, can be declared as "may unwind" via
+/// their ABI (e.g. `extern "C-unwind"`). To uphold the guarantees that
+/// Rust-defined functions never unwind a well-behaved Rust program needs to
+/// catch unwinding from foreign functions and force them to abort.
+///
+/// This pass walks walks over all functions calls which may possibly unwind,
+/// and if any are found sets their cleanup to a block that aborts the process.
+/// This forces all unwinds, in panic=abort mode happening in foreign code, to
+/// trigger a process abort.
```

Comment:
To me this sounds like a description of a soundness fix for https://github.com/rust-lang/rust/issues/83116. But your PR description makes it sound like this just reimplements an existing thing in a different way. Which of the two is true? I am quite confused here.^^

Reply:
Basically all of that is fixed by this pass. This pass:

* Fixes #83116 - there was no logic in the compiler before to insert an aborting edge on the unwind.
* Reimplements the previous abort-on-unwind logic. Previously this was a small tweak during MIR construction, but I backed out that part of MIR construction and instead this pass takes care of it now. This is also the behavior change I mentioned above, where previously the frame's destrutors would run before aborting and now they won't.